complete docs on corpora (#70)
fixie-ai/ultradox: api-reference/corpora/overview.mdx

---
title: "Corpus Service (RAG) Overview"

> description: "Understanding RAG in Ultravox"

Reviewer: ```suggestion
description: "Understanding Retrieval Augmented Generation in Ultravox"
```

---

> The Ultravox corpus service provides powers RAG for your agents. All of the relevant APIs appear under `/api/corpora` (note: 'corpora' is the plural of 'corpus').

Reviewer: ```suggestion
The Ultravox corpus service is the easiest way to add [RAG](https://en.wikipedia.org/wiki/Retrieval-augmented_generation) to your agents. All of the relevant APIs appear under `/api/corpora` (note: 'corpora' is the plural of 'corpus').
```

## Corpus Overview
```mermaid

> flowchart LR
>     1(Corpus) -->|contains| 2(Source) -->|contains| 3(Document) -->|contains| 4(Chunk) -->|contains| 5(Vector)

Reviewer: ```suggestion
flowchart LR
    1(Corpus) -->|contains| 2(Source) -->|produces| 3(Document) -->|comprised of| 4(Chunk) -->|referenceable by| 5(Vector)
```

```

> Each `Corpus` can contain one or more sources. Each `Source` can contain one or more documents. Each `Document` is broken up into one or more chunks. Each `Chunk` is then stored as one or more vectors.

Reviewer: ```suggestion
Each `Corpus` can contain one or more sources. Each `Source` contributes one or more documents, which are single files (HTML page, PDF, DOC, etc.). Each `Document` is broken up into one or more chunks, which are the units that may be returned for a query. For each `Chunk` several vectors are produced for similarity search.
```

The API exposes metadata about documents as well as the number of chunks and vectors in a corpus. However, there are no APIs provided to directly manipulate or edit documents, chunks, or vectors. If you need to update or delete documents, you must update or delete the source.

## Providing Corpus Access to Agents
<Warning>
  <b>Keep API Keys Safe</b>
  <br />
  Using the [Query Corpus](/api-reference/corpora/corpus-query) endpoint directly from your agents could expose your Ultravox API key. Use the `queryCorpus` tool instead.
</Warning>
We provide the built-in [queryCorpus](/essentials/tools#built-in-tools) tool that should be used by your agents. For example, if we wanted to create a voice agent to answer questions about Seattle, we could provide the tool like this:
    ```json
    {
      "systemPrompt": "Use the queryCorpus tool to answer questions about Seattle.",
      "selectedTools": [
        {
          "toolName": "queryCorpus", 
          "parameterOverrides": {
            "corpus_id": "<your_corpus_id_here>",
            "max_results": 5
          }
        }
      ]
    }
    ```

### queryCorpus Tool Performance
- Maximum of 20 chunks per query

>   - Model dynamically chooses value between 1-20
>   - Query response latency is not impacted by the number of chunks returned
>   - More chunks returned will grow the overall agent context and will mean slower subsequent generations

Reviewer: ```suggestion
  - By default, the model may dynamically choose a value between 1-20, defaulting to 5. A static value may be provided via override.
  - Query response latency is not significantly impacted by the number of chunks returned.
  - More chunks returned will grow the overall agent context, which will mean slower subsequent generations.
  - More chunks may also confuse the agent if most chunks are irrelevant.
```

## Limits and Configuration

### Account Limits
- Default limit is 2 corpora per account
- No limit on the number of sources within a corpus

### Document Management
- Documents can be viewed but not edited directly
- To update documents, create a new source

> - Supported document types:

Reviewer: Maybe some note about requests for other types?  Some wouldn't be hard to re-add.

Author: good idea. added link to discord.

  - Text files (including Markdown)
  - PDFs
  - Word documents
  - EPUB files
  - PowerPoint presentations

### Source Management

> - PUT request on a source triggers a refresh while maintaining the ID and created timestamp

Reviewer: We should probably steer toward PATCH, which is safer in general as it will only change the fields included in the request.  That said, any update to a source will result in a recrawl.  It's also probably worth mentioning that the corpus is still queryable during a recrawl, but the results obviously may not be consistent.

Author: ok done
